odoo/addons/base/res/res_partner.py

```diff
@@ -326,6 +326,19 @@ def copy(self, default=None):
         default = dict(default or {}, name=_('%s (copy)') % self.name)
         return super(Partner, self).copy(default)
 
+    @api.onchange('state_id')
+    def onchange_state_id(self):
+        if self.state_id:
+            self.country_id = self.env['res.country.state'].search([('id', '=', self.state_id.id)]).country_id.id
```

Comment:
No need to search on state. Just **self.country_id = self.state_id.country_id** is enough.